services/wireguard/service.go

```diff
@@ -0,0 +1,151 @@
+	return &Manager{
+		locationResolver: locationResolver,
+		ipResolver:       ipResolver,
+	}
+}
+
+// Manager represents entrypoint for Wireguard service
+type Manager struct {
+	process          sync.WaitGroup
+	locationResolver location.Resolver
+	ipResolver       ip.Resolver
+	isStarted        bool
+}
+
+// Config represent a Wireguard service provider configuration that will be passed to the consumer for establishing a connection
+type Config struct {
+	PublicKey string
+	IP        string
+	Endpoint  string
+}
+
+// Start starts service - does not block
+func (manager *Manager) Start(providerID identity.Identity) (dto_discovery.ServiceProposal, session.ConfigProvider, error) {
+	publicIP, err := manager.ipResolver.GetPublicIP()
+	if err != nil {
+		return dto_discovery.ServiceProposal{}, nil, err
+	}
+
+	sessionConfigProvider := func() (session.ServiceConfiguration, error) {
+		return setupWireguard(publicIP)
+	}
+
+	if manager.isStarted {
+		return dto_discovery.ServiceProposal{}, sessionConfigProvider, ErrAlreadyStarted
+	}
+
+	manager.process.Add(1)
+	manager.isStarted = true
+	log.Info(logPrefix, "Wireguard service started successfully")
+
+	country, err := manager.locationResolver.ResolveCountry(publicIP)
+	if err != nil {
+		return dto_discovery.ServiceProposal{}, nil, err
+	}
+
+	proposal := dto_discovery.ServiceProposal{
+		ServiceType: ServiceType,
+		ServiceDefinition: ServiceDefinition{
+			Location: dto_discovery.Location{Country: country},
+		},
+		PaymentMethodType: PaymentMethod,
+		PaymentMethod: Payment{
+			Price: money.NewMoney(0, money.CURRENCY_MYST),
+		},
+	}
+
+	return proposal, sessionConfigProvider, nil
+}
+
+// Wait blocks until service is stopped
+func (manager *Manager) Wait() error {
+	if !manager.isStarted {
+		return nil
+	}
+	manager.process.Wait()
+	return nil
+}
+
+// Stop stops service
+func (manager *Manager) Stop() error {
+	if !manager.isStarted {
+		return nil
+	}
+
+	manager.process.Done()
+	manager.isStarted = false
+	log.Info(logPrefix, "Wireguard service stopped")
+	return nil
+}
+
+func setupWireguard(publicIP string) (Config, error) {
+	// TODO initialize wireguard interface.
+	// out, err := exec.Command("ip", "link", "add", "dev", "wg0", "type", "wireguard").CombinedOutput()
+	// out, err := exec.Command("ip", "address", "add", "dev", "wg0", "192.168.100.1/24").CombinedOutput()
+
+	// TODO configure wireguard interface.
+	// client, err := wireguardctrl.New()
+	// if err != nil {
+	// 	return Config{}, err
+	// }
+
+	// TODO wireguard device configuration like private key, listen port, peer list should survive through restarts.
+	// TODO we need to have some persistent storage for it.
+	// client.ConfigureDevice("wg0", wgtypes.Config{
+	// 	PrivateKey: "8C6Pp0cI2tgLeYOMVbnMMLl/zN2blFK+OWGaIxX0bHY=",
+	// 	ListenPort: 52820,
+	// 	Peers:      nil,
+	// })
+
+	// TODO if the wireguard interface already configured we can get required parameters from it.
+	// device, err := client.Device("wg0")
```

Comment:
WG device name should be selected only after listing all active devices. Then choosing free number. We should know which device constitutes which session.